### DISCUSSION TITLE
팀 컨벤션

### DISCUSSION THREAD

**Discussion by moomint8** [AI-X-Jiwoo/Jiwoo#2] [Ground Rules]
# Git 작업 단위
### PR: 하나의 Entity에 대해 하나의 기능 완성 시 dev(Git flow) 또는 main(Github flow) 브랜치에 PR

# 구현 순서(TDD 기반)
- DB 설계 및 요구사항에 의거해 Entity 생성
- 요구사항을 구현할 Service Interface 생성
- Test Code 작성(given/when/then)  + **Git Pull** + commit
   - Test Code는 성공 케이스 외에도 여러 실패 / 예외 케이스에 대해 상세히 작성(음수, 존재하지 않는 값 등)
   - given/when/then 구조는 가독성도 향상 시킴 -> 유지 보수 향상, 협업 용이(개발 시간 단축)
- 실제 기능이 동작할 Service Implement 구현 + Test Code 통과 -> **Git Pull** + commit
- Controller 구현 + POSTMAN 또는 Mock 등을 이용해 테스트 + **Git Pull** + commit
- **Git Pull** + PR
- 요구 사항 변경 또는 추가 시 이에 맞게 각 Test Code, Service 등 개선
### 기능을 분리하는 부분이 헷갈리는 경우, 자신이 구현할 기능에 대해 명세하면 좀 더 용이해짐
![Untitled](https://github.com/user-attachments/assets/4d8a7584-79a3-4076-a2e6-4165d6888bea)

# Commit
### Commit Type
![Untitled (1)](https://github.com/user-attachments/assets/34f0a13a-b209-4deb-b8a1-144aacf32fe1)
### Commit 컨벤션
- 타입(대상): 구현 내용               // 제목, 짧게 과거형x 명령형으로 작성
   구현 상세 내용 및 설명            // 본문
- commit 메시지 의미있게 작성(메시지만으로 어떤 기능이 어떤 작업되었는지 파악할 수 있도록)
- commit 하기 전 **브랜치 반드시 확인**

# Code
- **메소드명, 변수명만으로 의도 파악 가능하도록 구현**
   - 주석 없이도 누구나 코드를 이해할 수 있도록 명명할 것
- **축약 금지**
   - 의도를 충분히 나타내기 위해 이름이 길어지는 것은 오히려 바람직함
   - 협의되지 않은 축약어는 유지 보수 저하, 협업에 어려움 발생(코테처럼 한 번 쓰고 버리는 코드가 아님을 명심)
- **공백도 컨벤션**
   - 적절히 사용된 공백은 문맥을 분리해줘 가독성을 매우 향상시킴
   - 줄바꿈 없이 작성된 글과 문단, 문장 분리가 잘 이루어진 글의 가독성을 생각하면 이해하기 용이
- **값 하드 코딩 금지**
   - 상수(final)를 활용할 것
- **변수 이름에 자료형 사용 지양**
   - 추후 자료형이 변경되는 경우 등의 경우 유지 보수에 어려움 유발
- **하나의 함수는 한 가지 기능만 담당(단일 책임의 원칙)**
   - 이를 잘 지킨 경우 유지 보수 용이 및 가독성 향상, 코드 재활용 용이
- **정상 케이스 뿐만 아니라 예외 케이스도 고민**
   - 사용자는 절대 우리 의도대로만 서비스를 이용하지 않음을 유의할 것

**Comment by Leegiyeon** [comment]
저는 공유 주신 내용대로 사용해도 좋습니다~
